### PR TITLE
[instance] Check the instance to be valid in platform callbacks

### PR DESCRIPTION
**Pull request by abtink**
This commit adds a new API `otInstanceIsInitialized()` to check if a
given instance is valid, i.e, it has been initialized and not
finalized. We then use this function to verify that passed-in instance
argument in platform callbacks is valid.

### CODE REVIEW

File: src/core/api/instance_api.cpp [hunk 1]
```diff
@@ -155,6 +155,12 @@ otInstance *otInstanceInit(void *aInstanceBuffer, size_t *aInstanceBufferSize)
     return instance;
 }
 
+bool otInstanceIsInitialized(otInstance *aInstance)
+{
+    return (aInstance != NULL);
+}
+
+
```
Comment (jwhui) [inline src/core/api/instance_api.cpp]: nit: extra newline

File: include/openthread/instance.h [hunk 2]
```diff
@@ -193,6 +193,17 @@ otInstance *otInstanceInit(void *aInstanceBuffer, size_t *aInstanceBufferSize);
 otInstance *otInstanceInitSingle(void);
 
 /**
+ * This function indicates whether or not the instance is valid, i.e., it has been initialized and not yet finalized.
+ *
```
Comment (jwhui) [inline include/openthread/instance.h]: Would be useful to expand this comment to indicate how the state of being initialized is actually determined.

Comment (abtink) [inline include/openthread/instance.h]: Added some more details in new commit...Let me know if this was what you had in mind...(feel free to suggest changed/edits... :) ...)

File: include/openthread/instance.h [hunk 3]
```diff
@@ -187,12 +187,28 @@ otInstance *otInstanceInit(void *aInstanceBuffer, size_t *aInstanceBufferSize);
...
+ * @retval A pointer to the single OpenThread instance structure.
  *
  */
 otInstance *otInstanceInitSingle(void);
 
 /**
+ * This function indicates whether or not the instance is valid/initialized.
+ *
+ * For single-instance case, the instance is considered valid if it is acquired and initialized using
+ * `otInstanceInitSingle()`. A sub-sequent call to `otInstanceFinalize()` causes the instance to be considered as
```
Comment (jwhui) [inline include/openthread/instance.h]: typo: `sub-sequent` -> `subsequent`

Comment (abtink) [inline include/openthread/instance.h]: Thanks... fixed it in the new commit.

**Pull request merged**